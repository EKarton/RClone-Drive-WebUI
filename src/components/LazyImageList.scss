.imagelist {
  height: 100%;

  &__row {
    display: flex;
    padding-bottom: 0.25rem;
  }
}
