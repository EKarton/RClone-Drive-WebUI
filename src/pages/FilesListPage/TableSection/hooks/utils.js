import { JobStatus, JobTypes } from 'utils/constants';

export function getFileNames(data) {
  return new Set(data.filter((file) => !file.IsDir).map((file) => file.Name));
}

export function getFolderNames(data) {
  return new Set(data.filter((file) => file.IsDir).map((file) => file.Name));
}

export function getUploadingJobs(jobs, remote, path) {
  return jobs.filter(
    (job) =>
      job.jobType === JobTypes.UPLOAD_FILE &&
      job.remote === remote &&
      job.dirPath.startsWith(path) &&
      job.status.value !== JobStatus.SUCCESS
  );
}

export function getUploadingFiles(jobs, remote, path) {
  return getUploadingJobs(jobs, remote, path).filter((job) => job.dirPath === path);
}

export function getUploadingFolders(jobs, remote, path) {
  return getUploadingJobs(jobs, remote, path).filter((job) => job.dirPath !== path);
}

export function getMovingJobs(jobs, remote, path) {
  const supportedJobTypes = new Set([
    JobTypes.MOVE_FILE,
    JobTypes.MOVE_FOLDER,
    JobTypes.RENAME_FILE,
    JobTypes.RENAME_FOLDER,
  ]);

  return jobs.filter((job) => {
    const correctJob = supportedJobTypes.has(job.jobType);
    const inProgress = job.status.value === JobStatus.ONGOING;

    if (!correctJob || !inProgress) {
      return false;
    }

    const isMatchSrc = job.src?.remote === remote && job.src?.dirPath === path;
    const isMatchTarget = job.target?.remote === remote && job.target?.dirPath === path;

    return correctJob && inProgress && (isMatchSrc || isMatchTarget);
  });
}
