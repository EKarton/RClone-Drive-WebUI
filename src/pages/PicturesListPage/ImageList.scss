.imagelist {
  &__row {
    display: flex;
    padding-bottom: 0.25rem;
  }
}
