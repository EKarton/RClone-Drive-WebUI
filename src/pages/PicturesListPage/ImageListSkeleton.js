import AutoSizer from 'react-virtualized-auto-sizer';
import { FixedSizeList } from 'react-window';
import './ImageListSkeleton.scss';
import { Skeleton } from '@mui/material';

export default function ImageListSkeleton() {
  const renderRow = (numImagesPerRow, style) => (
    <div className="imagelist-skeleton__row" style={style}>
      {Array.from({ length: numImagesPerRow }, () => (
        <Skeleton variant="rectangular" className="imagelist-skeleton__img" />
      ))}
    </div>
  );

  return (
    <AutoSizer>
      {({ height, width }) => {
        const numImagesPerRow = width < 1920 ? 3 : 5;
        const imgWidth = Math.floor(width / numImagesPerRow);
        const imgHeight = imgWidth;
        const numRows = Math.ceil(height / imgHeight);

        return (
          <FixedSizeList
            width={width}
            height={height}
            itemCount={numRows}
            itemSize={imgHeight}
          >
            {({ style }) => renderRow(numImagesPerRow, style)}
          </FixedSizeList>
        );
      }}
    </AutoSizer>
  );
}
