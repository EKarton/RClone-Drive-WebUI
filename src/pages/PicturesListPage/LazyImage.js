import { Skeleton } from "@mui/material";
import useRCloneClient from "hooks/useRCloneClient";
import { useEffect, useState } from "react";
import { useInView } from "react-intersection-observer";
import "./LazyImage.scss";

const LazyImage = ({ image, width, height }) => {
  const { ref, inView } = useInView({
    triggerOnce: true,
    rootMargin: "0px",
  });

  const rCloneClient = useRCloneClient();
  const [imageUrl, setImageUrl] = useState();
  const [error, setError] = useState();

  useEffect(() => {
    const fetchData = async () => {
      try {
        setError(undefined);
        setImageUrl(undefined);

        const imageBlob = await rCloneClient.fetchFileContents(
          image.remote,
          image.folderPath,
          image.fileName
        );

        const imageUrl = URL.createObjectURL(new Blob([imageBlob]));
        setImageUrl(imageUrl);
      } catch (error) {
        setError(error);
      }
    };

    if (inView && !imageUrl && image) {
      fetchData();
    }

    return () => {
      if (imageUrl) {
        URL.revokeObjectURL(imageUrl);
      }
    };
  }, [image, imageUrl, inView, rCloneClient]);

  const renderContent = () => {
    if (error) {
      return <div>Error! {error.message}</div>;
    }

    if (!imageUrl) {
      return (
        <Skeleton variant="rectangular" className="lazy-image__skeleton" />
      );
    }

    return (
      <img
        className="lazy-image__img"
        src={imageUrl}
        alt={image.fileName}
        loading="lazy"
        width={width}
        height={height}
      />
    );
  };

  return (
    <div ref={ref} data-inview={inView} style={{ width, height }}>
      {inView ? renderContent() : null}
    </div>
  );
};

export default LazyImage;
