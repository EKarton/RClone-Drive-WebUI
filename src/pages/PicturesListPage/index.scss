.pictures-list-page {
  &__header {
    padding-bottom: 1rem;
  }
}
