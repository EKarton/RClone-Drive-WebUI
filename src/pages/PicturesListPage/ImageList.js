import LazyImageList from 'components/LazyImageList';
import LazyImageListSkeleton from 'components/LazyImageListSkeleton';
import useFetchPictures from 'hooks/fetch-data/useFetchPictures';
import useFileViewerDialog from 'hooks/utils/useFileViewerDialog';
import useRecentlyViewedImages from 'hooks/utils/useRecentlyViewedImages';
import { StatusTypes } from 'utils/constants';

export default function ImageList({ remote, path }) {
  const fileViewer = useFileViewerDialog();
  const recentlyViewedImages = useRecentlyViewedImages();
  const { status, data, error } = useFetchPictures(remote, path);

  if (status === StatusTypes.LOADING) {
    return <LazyImageListSkeleton />;
  }

  if (status === StatusTypes.ERROR) {
    throw error;
  }

  const handleImageClicked = (image) => {
    recentlyViewedImages.addImage(image);
    fileViewer.show(image);
  };

  return (
    <LazyImageList images={data} remote={remote} onImageClicked={handleImageClicked} />
  );
}
