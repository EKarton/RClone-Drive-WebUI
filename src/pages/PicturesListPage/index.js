import Breadcrumbs from 'components/Breadcrumbs';
import { FileViewerDialogProvider } from 'contexts/FileViewerDialog';
import useRemotePathParams from 'hooks/utils/useRemotePathParams';
import PicturesListPageErrorBoundary from 'pages/ErrorBoundaries/PicturesListPageErrorBoundary';
import ImageList from './ImageList';
import './index.scss';

export default function PicturesListPage() {
  const { remote, path } = useRemotePathParams();

  return (
    <PicturesListPageErrorBoundary>
      <FileViewerDialogProvider>
        <Breadcrumbs
          remote={remote}
          path={path}
          homeText="My Pictures"
          homePath="/pictures"
          className="pictures-list-page__header"
        />
        <ImageList remote={remote} path={path} />
      </FileViewerDialogProvider>
    </PicturesListPageErrorBoundary>
  );
}
