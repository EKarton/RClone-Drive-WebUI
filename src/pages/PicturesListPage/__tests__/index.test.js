import { Routes, Route } from 'react-router';
import LazyImageList from 'components/LazyImageList';
import useFetchPictures from 'hooks/fetch-data/useFetchPictures';
import { StatusTypes } from 'utils/constants';
import { hashRemotePath } from 'utils/remote-paths-url';
import { mockPictures } from 'test-utils/mock-responses';
import { customRender } from 'test-utils/react';
import PicturesListPage from '../index';

jest.mock('hooks/fetch-data/useFetchPictures');
jest.mock('components/LazyImageList');

describe('PicturesListPage', () => {
  const remote = 'googledrive';

  beforeEach(() => {
    useFetchPictures.mockResolvedValue({
      status: StatusTypes.SUCCESS,
      data: mockPictures.list,
    });

    LazyImageList.mockReturnValue(null);
  });

  it('should not throw an error when the api call succeeds', async () => {
    const { baseElement } = renderComponent();

    expect(baseElement).toMatchSnapshot();
  });

  const renderComponent = () => {
    const route = `/pictures/${hashRemotePath(`${remote}:`)}`;

    const initialRCloneInfoState = {
      endpoint: 'http://localhost:5572',
      username: 'local',
      password: '1234',
    };

    const component = (
      <Routes>
        <Route path="/pictures/:id" element={<PicturesListPage />} />
      </Routes>
    );

    return customRender(component, { initialRCloneInfoState }, { route });
  };
});
