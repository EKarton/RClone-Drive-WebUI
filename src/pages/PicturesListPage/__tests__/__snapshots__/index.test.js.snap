// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PicturesListPage should not throw an error when the api call succeeds 1`] = `
<body>
  <div>
    <nav
      aria-label="breadcrumb"
      class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root pictures-list-page__header css-63vwl1-MuiTypography-root-MuiBreadcrumbs-root"
    >
      <ol
        class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
      >
        <li
          class="MuiBreadcrumbs-li"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1kgfegd-MuiTypography-root-MuiLink-root"
            href="/pictures"
          >
            My Pictures
          </a>
        </li>
        <li
          aria-hidden="true"
          class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
        >
          /
        </li>
        <li
          class="MuiBreadcrumbs-li"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1kgfegd-MuiTypography-root-MuiLink-root"
            href="/pictures/Z29vZ2xlZHJpdmU6"
          >
            googledrive
          </a>
        </li>
      </ol>
    </nav>
  </div>
</body>
`;
